infrastructure/openshift/retrieve-pnet-logs.sh

```diff
@@ -0,0 +1,59 @@
+#!/usr/bin/env bash
+
+# Exit immediately if a command exits with a non-zero status.
+# set -e
+
+function get_last_tx_id() {
+  # See https://www.linuxjournal.com/content/return-values-bash-functions
+  local  __resultvar=$1
+  local  tx_id=''
+
+  # Set to enter loop
+  local db_status=-1
+  while [ $db_status -ne 0 ]
+  do
+    echo -e "-------- get_last_tx_id calling psql --------\n"
+    tx_id=$(psql -h ${PGHOST} -d ${PGDATABASE} -U ${PGUSER} -t -c 'select max(ptl."TransactionId") from "PharmanetTransactionLog" ptl')
+    db_status=$?
+    echo -e "Last psql status:  _${db_status}_\n"
+    if [ $db_status -ne 0 ]; then sleep 10; fi
+  done
+
+  # Trim whitespace
+  tx_id=`echo ${tx_id} | sed 's/^ *//g'`
+  # Handle initial empty table condition
+  if [ "${tx_id}" = '' ]; then tx_id='0'; fi
+  # "Return" result to caller
+  eval $__resultvar="${tx_id}"
+}
+
+
+function main() {
+  echo -e "-------- STARTING CRON --------\n"
+
+  echo -e "Connecting to database host:  _${PGHOST}_\n"
+  echo -e "API client name:  _${PRIME_ODR_API_CLIENT_NAME}_\n"
+  echo -e "Fetch size:  _${PRIME_ODR_API_FETCH_SIZE}_\n"
+
+  HAS_MORE='Y'
+
+  while [ "${HAS_MORE}" = 'Y' ]
+  do
+    get_last_tx_id LAST_TX_ID
+    echo -e "Last transaction id:  _${LAST_TX_ID}_\n"
+
+    UUID=$(cat /proc/sys/kernel/random/uuid)
+    echo -e "Generated request id:  _${UUID}_\n"
+
+    # CA certs need to be in place:  https://stackoverflow.com/questions/3160909/how-do-i-deal-with-certificates-using-curl-while-trying-to-access-an-https-url
+    # ls -l /etc/ssl/certs
+    curl --cert /opt/certs/prime-odr-api-cert.crt:${PRIME_ODR_API_SSL_CERT_PASSWORD} --key /opt/certs/prime-odr-api-cert.key --header "Authorization: Basic ${PRIME_ODR_API_ENCODED_CREDENTIALS}" \
+      "https://t1primedatasvc.maximusbc.ca/odr/prime/pnetdata/transactionLog?requestUUID=${UUID}&clientName=${PRIME_ODR_API_CLIENT_NAME}&lastTxnId=${LAST_TX_ID}&fetchSize=${PRIME_ODR_API_FETCH_SIZE}" | \
+      python3 /opt/scripts/parse_api_response.py | \
```

Comment:
Does this script prevent sql injection etc?

Reply:
The Postgres COPY command seems to be doing something smarter than simply concatenating input for an INSERT INTO statement (e.g. 
```
insert into "PharmanetTransactionLog"("TransactionId", "TxDateTime", "UserId", "SourceIpAddress", "LocationIpAddress", "PharmacyId", "TransactionType", "TransactionSubType", "PractitionerId", "CollegePrefix", "TransactionOutcome", "ProviderSoftwareId", "ProviderSoftwareVersion") 
values(16, TO_TIMESTAMP('20210726T185028', 'YYYYMMDDTHH24MISS'), '923456', '010.085.013.001', '010.085.013.193', 'BC00000J70', 'PE', '02', '700005', 'P1', 'TPN', '', ''); DELETE FROM "PlrProvider"; --')
```
)

Reply:
I tried the following:
```
echo "64,20210726T185028.52,923456,
010.085.013.001,010.085.013.193,BC00000J70,PE,02,700005,P1,TPN,,'); DELETE FROM "PlrProvider"; --" |  psql -h localhost -p 5432
-U postgres -c "\copy \"PharmanetTransactionLog\"(\"TransactionId\", \"TxDateTime\", \"UserId\", \"SourceIpAddress\", \"Location
IpAddress\", \"PharmacyId\", \"TransactionType\", \"TransactionSubType\", \"PractitionerId\", \"CollegePrefix\", \"TransactionOu
tcome\", \"ProviderSoftwareId\", \"ProviderSoftwareVersion\") FROM STDIN (FORMAT CSV)"
```

and nothing was deleted from the `PlrProvider` table and the following was inserted:

![image](https://user-images.githubusercontent.com/7586665/132888248-69238b2e-ed78-4897-ae2c-ff00a63e9bff.png)

so the Postgres COPY command seems to mitigate SQL injection.